src/transformers/models/idefics3/image_processing_idefics3.py

```diff
@@ -0,0 +1,847 @@
+
+    Returns:
+        The output size of the image after resizing.
+    """
+    if resolution_max_side > max_image_size:
+        raise ValueError("`resolution_max_side` cannot be larger than `max_image_size`")
+
+    if isinstance(image, Image.Image):
+        width, height = image.size
+    else:
+        height, width = get_image_size(image, channel_dim=input_data_format)
+
+    # Find the output size, when rescaling the longest edge to max_len and preserving the aspect ratio
+    height, width = _resize_output_size_rescale_to_max_len(
+        height, width, max_len=resolution_max_side
+    )
+    # Find the output size when scaling the image to be below the max_image_size
+    height, width = _resize_output_size_scale_below_upper_bound(
+        height, width, max_len=max_image_size
+    )
+    return height, width
+
+
+def split_image(
+    image: np.ndarray,
+    max_image_size: Dict[str, int],
+    resample: PILImageResampling = PILImageResampling.LANCZOS,
+    input_data_format: Optional[Union[str, ChannelDimension]] = None,
+    data_format: Optional[Union[str, ChannelDimension]] = None,
+):
+    """
+    Image splitting strategy.
+    1) If one side of the original image is larger than `max_image_size`, resize it to `max_image_size` while preserving the aspect ratio.
+    2) Divide the resulting image into `ceil(height / max_image_size)` x `ceil(width / max_image_size)`
+    sub-images of approximately the same size each (up to the fact that `vision_encoder_max_image_size` does not divide `height` or
+    `width`).
+    3) Returns the list of the crops and the original image, in addition to the number of splits for the height and the width.
+    """
+    if isinstance(image, Image.Image):
+        width, height = image.size
+    else:
+        height, width = get_image_size(image, channel_dim=input_data_format)
+    max_height = max_width = max_image_size["longest_edge"]
+
+    frames = []
+    if height > max_height or width > max_width:
+        # Calculate the number of splits
+        num_splits_h = math.ceil(height / max_height)
+        num_splits_w = math.ceil(width / max_width)
+        # Calculate the optimal width and height for the sub-images
+        optimal_height = math.ceil(height / num_splits_h)
+        optimal_width = math.ceil(width / num_splits_w)
+
+        # Iterate through each row and column
+        for r in range(num_splits_h):
+            for c in range(num_splits_w):
+                # Calculate the starting point of the crop
+                start_x = c * optimal_width
+                start_y = r * optimal_height
+
+                # Calculate the ending point of the crop
+                end_x = min(start_x + optimal_width, width)
+                end_y = min(start_y + optimal_height, height)
+
+                # Crop the image
+                if isinstance(image, Image.Image):
+                    cropped_image = image.crop((start_x, start_y, end_x, end_y))
```

Comment:
Same here - there shouldn't be any PIL code for the transformations

Reply:
After further discussion with Amy, I added a large change to support processing the images as numpy arrays. Here: https://github.com/huggingface/transformers/pull/32473/commits/5a0c0f4f89d783ccbe2b57cf914bf79d1a083078